runtime/src/message_processor.rs

```diff
@@ -237,12 +326,49 @@ impl MessageProcessor {
                 &native_loader::id(),
                 &keyed_accounts,
                 &instruction.data,
+                invoke_context,
             )
         } else {
             Err(InstructionError::UnsupportedProgramId)
         }
     }
 
+    /// Process a cross-program instruction
+    /// This method calls the instruction's program entrypoint method
+    pub fn process_cross_program_instruction(
+        message: &Message,
+        executable_accounts: &[(Pubkey, RefCell<Account>)],
+        accounts: &[Rc<RefCell<Account>>],
+        signers: &[Pubkey],
+        process_instruction: ProcessInstructionWithContext,
+        invoke_context: &mut dyn InvokeContext,
+    ) -> Result<(), InstructionError> {
+        let instruction = &message.instructions[0];
+
+        // Verify the calling program hasn't misbehaved
+        invoke_context.verify_and_update(message, instruction, signers, accounts)?;
+
+        // Construct keyed accounts
+        let keyed_accounts =
+            Self::create_keyed_accounts(message, instruction, executable_accounts, accounts)?;
+
+        // Invoke callee
+        invoke_context.push(instruction.program_id(&message.account_keys))?;
+        let mut result = process_instruction(
+            &keyed_accounts[0].owner()?,
+            &keyed_accounts,
+            &instruction.data,
+            invoke_context,
+        );
+        if result.is_ok() {
+            // Verify the called program has not misbehaved
+            result = invoke_context.verify_and_update(message, instruction, signers, accounts);
+        }
+        invoke_context.pop();
```

Comment:
@jackcmay Sorry for being after-party comment, but I just started to looking the program-related code. :)

This `invoke_context.pop()` isn't always be paired with `invoke_context.push(...)` when [`keyed_accounts[0].owner()?`](https://github.com/solana-labs/solana/pull/9582/files#diff-f3f8bbf10039e77655b10f56bc48b473R358) fails to borrow. Is that ok?

And qq: is cross-program invocation can be gracefully handled inside a invoker program in some way when invokee's result is `InstructionError`?
Then, [this verification](https://github.com/solana-labs/solana/pull/9582/files#diff-f3f8bbf10039e77655b10f56bc48b473R219) might start to reference a wrong `program_id`.

I think to it's better to avoid these kind of things like with closure surrounded by `push()` and `pop()`

Reply:
Yeah, that will never fail in this implementation so I left it.  I recently updated this code to extend cross-program invocation to native programs and in those changes the case you mention no longer exists.

I'm not quite following the "qq", if the invoke fails the result is held and context poped before returning.  Can you elaborate?